### PR TITLE
loader: do ELF substitutions in memory

### PR DESCRIPTION
**Pull request by lmb**
pkg/bpf: allow overriding constants during load

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: remove loader.ELFSubstitutions

    Seems like a needless abstraction, remove it.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: make EndpointHash infallible

    The loader exports a method to turn an endpoint into a hash which is meant
    to change any time the BPF needs to be regenerated. This method is currently
    fallible because of a work around needed because hash.Hash is not easily
    copied. This in turn causes a bunch of awkward error handling where there is
    none needed. It also means that the endpoint code in theory has to deal with 
    the case of datapathRegenCtx.bpfHeaderfileHash being empty.

    Replace the fallible implementation with an infallible one. The new
    implementation doesn't produce the same hashes as the old one, but this
    doesn't matter since the semantics are unchanged. The hashes are used to
    identify templates which aren't reused across agent restarts.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: stop linking template.o into ep.StateDir()

    Compilation of an endpoint program creates a template.o symlink in the
    endpoint state directory. This is problematic since we really should be
    creating the file in the "next" state directory. Overall, the loader should
    not fiddle around in the endpoints state directory.

    Replace the symlink with template.txt in the endpoint state directory, which
    contains the hash of the used template. Importantly, we write this file from
    the endpoint code, not from the loader.

    Move the only user of template.o to template.txt.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: remove ELF substitution logic

    In theory, we need to recompile the BPF for an endpoint any time its
    configuration changes. Compiling is fairly expensive, so a mechanism called
    ELF templating / substitution was developed to avoid recompilation in the
    most common cases.

    1. bpf_lxc.c or bpf_host.c are compiled with a "generic"
      ep_config.h that contains placeholders into
      template.o.

    2. For each endpoint, we replace all the placeholders in
      template.o and write a bpf_lxc.o, etc.

    3. Each time we need to reload the datapath bpf_lxc.o is read
      from disk.

    Replacing the placeholders in step (2) works by parsing the ELF and
    manipulating the string table and data sections directly.

    I believe that this approach was chosen when the loader still relied on
    iproute2 to do all the loading. Now we use cilium/ebpf and the templating
    mechanism is becoming a problem. It is very hard to follow how it works and
    it impedes other refactoring.

    Replace ELF templating with manipulation of ebpf.CollectionSpec directly.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

remove pkg/elf

    There are no users of the package left in the code base. Remove it.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

```release-note
loader: Significantly reduce memory usage during endpoint regeneration
```

### CODE REVIEW

**Review by dylandreimerink — APPROVED** [review]
:+1:

**Review by derailed — APPROVED** [review]
@lmb Nice work and clean up!

File: pkg/bpf/collection.go [hunk 1]
```diff
@@ -390,11 +413,17 @@ func inlineGlobalData(spec *ebpf.CollectionSpec) error {
...
-			value, ok := vars[off]
+			v, ok := vars[off]
 			if !ok {
 				return fmt.Errorf("no global constant found in %s at offset %d", globalDataMap, off)
 			}
 
+			value := v.value
+			name := strings.TrimPrefix(v.name, constantPrefix)
+			if override, ok := overrides[name]; ok {
+				value = override
```
Comment (ti-mo) [inline pkg/bpf/collection.go]: Can this be done earlier in the function, for example in `for name := range overrides`? I'd merge overrides into `vars` to avoid having this logic here, see my comment below.

File: pkg/bpf/collection.go [hunk 2]
```diff
@@ -464,15 +499,16 @@ func globalData(spec *ebpf.CollectionSpec) (varOffsets, error) {
...
 			value = uint64(buf[0])
 		default:
-			return nil, fmt.Errorf("invalid variable size %d", vsi.Size)
+			return nil, nil, fmt.Errorf("invalid variable size %d", vsi.Size)
 		}
 
 		// Emit the variable's value by its offset in the datasec.
-		out[vsi.Offset] = value
+		out[vsi.Offset] = globalVar{v.Name, value}
+		names[v.Name] = struct{}{}
```
Comment (ti-mo) [inline pkg/bpf/collection.go]: We could use names as a two-way index:
```suggestion
		names[v.Name] = vsi.Offset
```
Then, in `inlineGlobalData`, we can use this to look up vars by name and merge overrides into values.

File: pkg/bpf/collection_test.go [hunk 3]
```diff
@@ -61,24 +63,24 @@ func TestLoadCollectionResizeLogBuffer(t *testing.T) {
...
+						{Offset: 4, Size: 2, Type: &btf.Var{Name: constantPrefix + "b"}},
+						{Offset: 8, Size: 8, Type: &btf.Var{Name: constantPrefix + "c"}},
 					},
 				},
 				Contents: []ebpf.MapKV{{Value: []byte{
-					// var 1
+					// var a
 					0x0, 0x0, 0x0, 0x80,
-					// var 2 has padding since var 3 aligns to 64 bits. Fill the padding
+					// var b has padding since var 3 aligns to 64 bits. Fill the padding
```
Comment (ti-mo) [inline pkg/bpf/collection_test.go]:
```suggestion
					// var b has padding since var c aligns to 64 bits. Fill the padding
```

File: pkg/bpf/collection_test.go [hunk 4]
```diff
@@ -117,9 +120,19 @@ func TestInlineGlobalData(t *testing.T) {
...
 		t.Errorf("unexpected Instruction constant: want: 0x%x, got: 0x%x", want, got)
 	}
+
+	spec = orig.Copy()
+	if err := inlineGlobalData(spec, map[string]uint64{"bogus": 42}); err == nil {
+		t.Fatal("Expected error when trying to replace bogus constant")
+	}
+
+	delete(spec.Maps, globalDataMap)
+	if err := inlineGlobalData(spec, map[string]uint64{"bogus": 42}); err != nil {
```
Comment (ti-mo) [inline pkg/bpf/collection_test.go]: Should we leave this nil to isolate the error condition better?
```suggestion
	if err := inlineGlobalData(spec, nil}); err != nil {
```

File: pkg/datapath/loader/netlink.go [hunk 5]
```diff
@@ -74,21 +90,18 @@ func replaceDatapath(ctx context.Context, opts replaceDatapathOptions) (_ func()
 		return nil, errors.New("opts.linkDir not set in replaceDatapath")
 	}
 
+	if opts.elf != "" {
+		return nil, errors.New("opts.elf is set")
```
Comment (ti-mo) [inline pkg/datapath/loader/netlink.go:94]:
```suggestion
		return nil, errors.New("opts.elf is not set")
```

**Review by ti-mo — APPROVED** [review]
:rocket: Left a few nits

File: pkg/datapath/loader/netlink.go (hunk 5 above)
Comment (lmb) [reply pkg/datapath/loader/netlink.go:94]: No, the error is correct.

File: pkg/bpf/collection.go (hunk 1 above)
Comment (lmb) [reply pkg/bpf/collection.go]: This is nice, thanks!

File: pkg/datapath/loader/netlink.go (hunk 5 above)
Comment (ti-mo) [reply pkg/datapath/loader/netlink.go:94]: Oh, I see..

**Comment by lmb** [comment]
@christarazi @qmonnet could you take a look please? This is finally ready!

**Review by christarazi — APPROVED** [review]
ci-structure LGTM

**Review by aanm — APPROVED** [review]
